File: ml-benchmark/autolifter/dac/msp.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = -4;
Config SampleIntMax = 4;
Config SampleSize = 8;
Config ComposeNum = 4;
Config NonLinear = true;

Inductive List = cons {Int, List} | nil Unit;

length = fix (\f: List->Int. \x: List.
  match x with
    cons {h, t} -> + (f t) 1
  | nil _ -> 0
  end
);

concat = fix (\f: List->List->List. \x: List. \y: List.
  match x with
    cons {h, t} -> cons {h, f t y}
  | nil _ -> y
  end
);

split = \xs: List. fix (\f: List->Int->{List, List}. \x: List. \n: Int.
  if (< n 1) then {nil unit, x}
  else match x with
    cons {h, t} -> let res = f t (- n 1) in {cons {h, res.1}, res.2}
  | _ -> {nil unit, nil unit}
  end
) xs (/ (length xs) 2);

dac = \v: List -> Int. \xs: List.
  let run = fix (\f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {_, nil _} -> xs
    | _ -> let sp = split xs in
      concat (f sp.1) (f sp.2)
    end
  ) in v (run xs);

/*User provided programs*/

max = \a: Int. \b: Int. if < a b then b else a;

msp =
  let mpp = fix (lambda f: Int -> List -> Int. \pre: Int. lambda l: List.
    match l with
      nil _ -> pre
    | cons {h, t} ->
      max pre (f (* h pre) t)
    end
  ) 1 in
  fix (\f: List -> Int. \xs: List.
  match xs with
    nil _ -> 1
  | cons {h, t} -> max (mpp xs) (f t)
  end
);

main = dac msp;

File: ml-benchmark/autolifter/dac/msp.ml
type l = Nil | Cons of int * l

let rec length x =
  match x with
  | Cons (h, t) -> length t + 1
  | Nil -> 0

let rec concat x y =
  match x with
  | Cons (h, t) -> Cons (h, concat t y)
  | Nil -> y

let split xs =
  let rec f x n =
    if n < 1 then (Nil, x)
    else
      match x with
      | Cons (h, t) ->
          let res = f t (n - 1) in
          match res with
          | (a, b) -> (Cons (h, a), b)
      | _ -> (Nil, Nil)
  in
  f xs (length xs / 2)

let dac v xs =
  let rec run xs =
    match xs with
    | Nil -> xs
    | Cons (h, t) ->
        match t with
        | Nil -> xs
        | _ ->
            let sp = split xs in
            match sp with
            | (a, b) -> concat (run a) (run b)
  in
  v (run xs)

let max a b = if a < b then b else a

let msp =
  let rec f pre l =
    match l with
    | Nil -> pre
    | Cons (h, t) -> max pre (f (h * pre) t)
  in
  let mpp = f 1 in
  let rec g xs =
    match xs with
    | Nil -> 1
    | Cons (h, t) -> max (mpp xs) (g t)
  in
  g

let program = dac msp